stl/src/cthread.cpp

```diff
@@ -99,9 +99,16 @@ _Thrd_id_t _Thrd_id() { // return unique id for current thread
 }
 
 unsigned int _Thrd_hardware_concurrency() { // return number of processors
-    SYSTEM_INFO info;
-    GetNativeSystemInfo(&info);
-    return info.dwNumberOfProcessors;
+    DWORD_PTR process_affinity;
+    DWORD_PTR system_affinity;
+    if (GetProcessAffinityMask(GetCurrentProcess(), &process_affinity, &system_affinity))

```

Comment:
[`GetProcessAffinityMask` documentation](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getprocessaffinitymask) says: 
> On a system with more than 64 processors, if the threads of the calling process are in a single processor group, the function sets the variables pointed to by `lpProcessAffinityMask` and `lpSystemAffinityMask` to the process affinity mask and the processor mask of active logical processors for that group. If the calling process contains threads in multiple groups, the function returns zero for both affinity masks.

[`dwNumberOfProcessors` documentation](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-system_info) says:
> The number of logical processors in the current group.

Is there a behavioral difference here? If so, is it desirable? (These are not rhetorical questions; I am far from a WinAPI expert.)

Reply:
Counting `GetProcessAffinityMask` is documented to return zero affinity for cases when process belongs to multiple processor groups.

As `dwNumberOfProcessors` documentation does not say anything about cases when process belongs to multiple processor groups. It could be assumed, that it still _the number of logical processors in the current group_.  (It might be however lack of documentation for such cases).

So there's no difference regarding maximum return until process belongs to multiple processor groups.
And there is a possible difference, when process belongs to multiple processor groups.

Process would belong to multiple processor groups after one of its thread is explicitly specified to run in different processor group (by a call to `SetThreadGroupAffinity` or as creation parameter). This is done manually for each thread, and `std::thread` will not do this by default. So zero result of `std::hardware_concurrency()` is legitimate for this case (the value is not well defined, it depends on users further calls to `SetThreadGroupAffinity`). Per-group value returned by `dwNumberOfProcessors`, which is supposedly 64 for multi-group cases, if it really still returns nonzero for such cases (system will try to pack 64 processors in one group) -- is also suitable, as `std::hardware_concurrency()` is a hint. Once user took control by `SetThreadGroupAffinity` call, they took responsibility about effective hardware concurrency.

This change primarily targets other case, when this or other process calls `SetProcessAffinityMask` for this process. This technique can be used to test concurrent algorithms for performance on fewer CPUs (remove some other cores, other SMTs or both), or for fixing concurrency bugs (Application Compatibility can do this, for example).

For multi-group cases - I'm not sure. There is a possibility to enumerate all processor groups by obtaining `GROUP_RELATIONSHIP` from `GetLogicalProcessorInformationEx`. But I do not know if this is the right thing to do, as bigger concurrency value obtained by such means is only available for those who will call `SetThreadGroupAffinity`.

Reply:
>And there is a possible difference, when process belongs to multiple processor groups.

I missed that. Sadly to test this we need a machine with more than 64 threads...

>(the value is not well defined, it depends on users further calls to SetThreadGroupAffinity).

I disagree, the value is defined for all portable uses of `std::thread`, since it uses the default group.

>There is a possibility to enumerate all processor groups

Right, we should not do that because `std::thread` can't reach the other groups.